core/sr-primitives/src/transaction_validity.rs

```diff
@@ -30,23 +29,240 @@ pub type TransactionLongevity = u64;
+	CannotLookup,
+	/// No validator found for the given unsigned transaction.
+	NoUnsignedValidator,
+	/// Any other custom unknown validity that is not covered by this enum.
+	Custom(u8),
+}
+
+impl Into<&'static str> for UnknownTransaction {
+	fn into(self) -> &'static str {
+		match self {
+			UnknownTransaction::CannotLookup =>
+				"Could not lookup some information to validate the transaction",
+			UnknownTransaction::NoUnsignedValidator =>
+				"Could not find an unsigned validator for the unsigned transaction",
+			UnknownTransaction::Custom(_) => "UnknownTransaction custom error",
+		}
+	}
+}
+
+/// The error that can occur while checking the validity of a transaction.
+#[derive(Clone, PartialEq, Eq, Encode, Decode, Copy)]
+#[cfg_attr(feature = "std", derive(Debug, serde::Serialize))]
+pub enum TransactionValidityError {
+	/// The transaction is invalid.
+	Invalid(InvalidTransaction),
+	/// Transaction validity can't be determined.
+	Unknown(UnknownTransaction),
+}
+
+impl TransactionValidityError {
+	/// Returns if the reason for the error was block resource exhaustion.
+	pub fn exhausted_resources(&self) -> bool {
+		match self {
+			Self::Invalid(e) => e.exhausted_resources(),
+			Self::Unknown(_) => false,
+		}
+	}
+}
+
+impl Into<&'static str> for TransactionValidityError {
+	fn into(self) -> &'static str {
+		match self {
+			TransactionValidityError::Invalid(invalid) => invalid.into(),
+			TransactionValidityError::Unknown(unknown) => unknown.into(),
+		}
+	}
+}
+
+impl From<InvalidTransaction> for TransactionValidityError {
+	fn from(err: InvalidTransaction) -> Self {
+		TransactionValidityError::Invalid(err)
+	}
+}
+
+impl From<UnknownTransaction> for TransactionValidityError {
+	fn from(err: UnknownTransaction) -> Self {
+		TransactionValidityError::Unknown(err)
+	}
+}
+
+impl Into<crate::ApplyError> for InvalidTransaction {
+	fn into(self) -> crate::ApplyError {
+		TransactionValidityError::from(self).into()
+	}
+}
+
+impl Into<crate::ApplyError> for UnknownTransaction {
+	fn into(self) -> crate::ApplyError {
+		TransactionValidityError::from(self).into()
+	}
+}
+
 /// Information on a transaction's validity and, if valid, on how it relates to other transactions.
 #[derive(Clone, PartialEq, Eq, Encode)]
 #[cfg_attr(feature = "std", derive(Debug))]
 pub enum TransactionValidity {
-	/// Transaction is invalid. Details are described by the error code.
-	Invalid(i8),
+	/// Transaction is invalid.
+	Invalid(InvalidTransaction),
 	/// Transaction is valid.
 	Valid(ValidTransaction),
 	/// Transaction validity can't be determined.
-	Unknown(i8),
+	Unknown(UnknownTransaction),
 }
 
-impl From<Result<ValidTransaction, DispatchError>> for TransactionValidity {
-	fn from(r: Result<ValidTransaction, DispatchError>) -> Self {
-		match r {
-			Ok(v) => TransactionValidity::Valid(v),
-			Err(e) => TransactionValidity::Invalid(e.into()),
+impl Into<TransactionValidity> for TransactionValidityError {
```

Comment:
Do we have `derive_more` around? We could get rid of all those `Into/From` stuff.